build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs

```diff
@@ -134,11 +146,59 @@ public IEnumerable<string> ListBlobs(string virtualDirectory)
                 }
             }
 
+            ResetLeaseRenewalTaskState();
+
             throw new Exception($"Unable to acquire lease on {blob}");
         }
 
+        private void ResetLeaseRenewalTaskState()
+        {
+            // Cancel the lease renewal task if it was created
+            if (_leaseRenewalTask != null)
+            {
+                _cancellationTokenSource.Cancel();
+                _leaseRenewalTask = null;
+            }
+        }
+
+        private static void AutoRenewLease(AzurePublisher instance, string blob, string leaseId)
+        {
+            // We will renew the lease every 45 seconds
+            TimeSpan maxWait = TimeSpan.FromSeconds(45);
+            TimeSpan delay = TimeSpan.FromMilliseconds(500);
+            TimeSpan waitFor = maxWait;
+
+            CancellationToken token = instance._cancellationTokenSource.Token;
+            while (true)
+            {
+                // If the task has been requested to be cancelled, then do so.
+                token.ThrowIfCancellationRequested();
+
+                try
+                {
+                    CloudBlockBlob cloudBlob = instance._blobContainer.GetBlockBlobReference(blob);
+                    AccessCondition ac = new AccessCondition() { LeaseId = leaseId };
+                    cloudBlob.RenewLeaseAsync(ac).Wait();
+                    waitFor = maxWait;
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine($"Retrying lease renewal on {blob}, {e.Message}");
+                    waitFor = delay;
+                }
+
+                // If the task has been requested to be cancelled, then do so.
+                token.ThrowIfCancellationRequested();
+
+                Thread.Sleep(waitFor);
+            }
+        }
+
         public void ReleaseLeaseOnBlob(string blob, string leaseId)
         {
+            // Cancel the lease renewal task since we are about to release the lease.
+            ResetLeaseRenewalTaskState();
+
```

Comment:
The way I see it, pre-change `( blob, leaseId )` forms a "lease" and there's no state in `AzurePublisher`. With this change, there's lease state in the `AzurePublisher` in a way that prevents two leases at the same time.

In any case... yeah, not important, since we don't currently need more than one lease.